Shape_detection/doc/Shape_detection/Shape_detection.txt

```diff
@@ -0,0 +1,596 @@
+
+The concept `NeighborQuery` provides the means for accessing neighbors of an item. 
+To create a model that respects this concept, the user has to provide an overload of the operator:
+
+- `NeighborQuery::operator()()` that has to fill a vector with indices of all items, which 
+are neighbors of the query item.
+
+
+\subsubsection Shape_detection_RegionGrowingFramework_conditions Regions
+
+The concept `RegionType` provides the means for validating regions. In fact, a model
+of this concept maintains a description of the region type that is used in region growing.
+To create a model that respect this concept, three functions have to be defined:
+
+- `RegionType::is_part_of_region()` This function checks if an item satisfies
+all necessary region requirements and can be added to a region. It is called per item.
+
+- `RegionType::is_valid_region()` This function checks if a region satisfies 
+all necessary region requirements. It is called per region.
+
+- `RegionType::update()` This utility function  enables to update any information, 
+which is maintained with the region.
+
+
+\subsubsection Shape_detection_RegionGrowingFramework_seeding Seeding
+
+The `SeedMap` property map, provided as an optional parameter to the main class, enables
+to define the seeding order of items that is which items are used first to grow regions from.
+Such items are referred to as *seed* items. The `SeedMap` maps the index of an item to its order 
+number in the overall region growing processing queue. The default map is the identity one 
+that is the seed index of the item equals to the item's index in the `input_range`. 
+If it maps to `std::size_t(-1)`, then the corresponding item is skipped.
+
+
+\subsubsection Shape_detection_RegionGrowingFramework_examples Examples
+
+This toy example shows how to define one's own \ref Shape_detection_RegionGrowingFramework_connectivity "NeighborQuery" and 
+\ref Shape_detection_RegionGrowingFramework_conditions "RegionType" classes, which are used to parameterize the 
+`CGAL::Shape_detection::Region_growing`. It also shows how to skip unnecessary items and change their default seeding order.
+
+We choose a simple custom object item. We define four such objects, where for each object, we manually 
+store indices of its neighbors. The operator `NeighborQuery::operator()()` does nothing but accessing these neighbors. 
+The `RegionType` class defines the three necessary functions:
+
+- `RegionType::is_part_of_region()` - `true` if the first and second objects are neighbors,
+- `RegionType::is_valid_region()` - always `true` after the first call to the function `update()`,
+- `RegionType::update()` - updates the internal flag from the default `false` to `true`.
+
+We also define a `SeedMap`, such that the second object is handled first, while the first object follows.
+Moreover, the last object is always skipped. Notice that in this example, the container with objects is `std::list`,
```

Comment:
> Notice that in this example, the container with objects is `std::list`

That's not the case `region_growing_with_custom_classes.cpp` uses a vector: `std::vector<std::size_t> neighbors;`

Yet the code traverses the vector as if it was a linked list:

```c++
    void operator()(
      const std::size_t query_index, 
      std::vector<std::size_t>& neighbors) const {
      
      std::size_t i = 0;
      for (const auto& object : m_objects) {
        if (i == query_index) {
          
          neighbors = object.neighbors;
          return;
        } ++i;
      }
    }
```

That is rather confusing.

Perhaps it once was a list but then the type was changed to a vector before the PR was submitted/merged, without updating the algorithm and docs?

Reply:
@nh2 Hi. You are completely right! Thank you for this remark! It is my fault. Indeed, I first used std::list and then updated to std::vector, because some reviewers did not like using the list, but I forgot to update the docs for some reason. I will fix it. However, note that RG can still work with std::list. In general, std::vector is not required, it is used only because it is more efficient.

Reply:
@danston Makes sense!

Another quick question:

From the added docs, on region growing, it doesn't seem to become quite clear how `max_accepted_angle` works. The docs say:

> `angle_threshold` - the maximum accepted angle between the normal associated with a point and the normal of a line/plane;

This suggests that the angle is indeed between each individual face and the fitted plane. However, I have observed the algorithm to produce two things that seem impossible under that interpretation, using face-based plane fitting:

* With an `angle_threshold` of 90: When I project associated faces onto the fitted plane, some project normals have opposite signum to the plane's normal
* With an `angle_threshold` of 70: Two noisy parallel planes facing each other being detected as part of the same region (shouldn't that be impossible, given that their face normals pretty much look at each other?)

I haven't fully read all implementation yet; maybe you can give me a quick hint at whether that's intended, and how we could make the documentation best reflect what's happening?

Reply:
I think I found the reason for the above:

In https://github.com/CGAL/cgal/blob/29f973a61628a14cc4ff7741cd3482e5eac8771c/Shape_detection/include/CGAL/Shape_detection/Region_growing/Region_growing_on_polygon_mesh/Least_squares_plane_fit_region.h#L205-L212

we have `abs()` which throws away the sign of the normal, thus also permitting faces that are within `180 <= here < 180-angle_threshold` degrees!

E.g. assume your fit plane's normal points exactly upwards. If you set the `angle_threshold` to, say, 5 degrees, this will not permit angles that are "pretty much exactly upwards" but also "pretty much exactly downwards".

Is that intended?

Reply:
I noticed another problem: The plane's normal is not estimated correctly (it is arbitrary, likely depending on the position of the plane relative to the origin).

I've put a fix for the above 2 issues into PR #4563.